### DISCUSSION TITLE
How to remove dependencies of package being removed?

### DISCUSSION THREAD

**Discussion by IIFE** [microsoft/vcpkg#18350] [Ideas]
Hi,

I have installed cpprestsdk, which itself depends on a number of boost packages that get installed as a result.

When I remove cpprestsdk, I wish to also remove all of its dependencies (i.e. boost).

I have looked into the `--recurse` option, but that does nothing.

Looking at the code, it doesn't appear that the remove command attempts to remove dependencies of the package being removed. What appears to be happening is that, when removing a package, any packages that have a dependencies on the package being removed are the ones that get removed. This is the opposite behaviour. The documentation for the `--recurse` options appears to suggest that dependencies of the package being removed are the ones that get removed, but looking at the implementation it seems to be the other way around.

Specifically, if we look at [this function](https://github.com/microsoft/vcpkg/blob/4da95d667c8600c80b4b5731631ebb7bcbc91f1b/toolsrc/src/vcpkg/dependencies.cpp#L452), you can see that only packages that depend on the package being removed are considered for removal, not the other way around.

Is this intended behaviour?

Thanks.

**Comment by PhoebeHui** [comment]
@IIFE, this is intended behavior,  if you remove boost, then adding --recurse option will remove all ports(eg: cpprestsdk)that dependent to boost.

**Comment by IIFE** [comment]
@PhoebeHui  In that case I don't think the `--recurse` documentation is clear. It doesn't really explain the actual purpose.

Additionally, is there a reason no option is available to remove dependencies of the package I wish to remove? I'm happy to look into it as I think it's a useful feature.

Thanks.

**Comment by invy** [comment]
Indeed, there should be the option (which is pretty easy to implement as it the same 'plan' which is created during export), but just remoe dependent packages.

**Comment by IIFE** [comment]
@invy Yes exactly what I had in mind. Extend [this function](https://github.com/microsoft/vcpkg/blob/4da95d667c8600c80b4b5731631ebb7bcbc91f1b/toolsrc/src/vcpkg/dependencies.cpp#L452) to return the dependencies of the package being removed.

Was thinking on the lines of passing a new flag to `create_remove_plan` to tell it whether to return the package dependencies, or retain current behaviour.

I'll create a PR if this sounds good to people. Happy to discuss alternatives.

Thanks.

**Comment by lazyhamster** [comment]
Don't forget you'll also have to check all other installed packages. Dependencies for currently removed package can also serve as dependencies for some other ports.

**Comment by vicroms** [comment]
Hi @IIFE 

Sorry we took so long to reply. If anyone in the community is interested in implementing it, we would gladly accept a PR to add said feature.

Some considerations that need to be made:

* Needs to be a new flag, e.g.: (`--upstream`, `--with-dependencies`, or something similar).
* Fail when attempting to delete a package that is a dependency of another package, that is not explicitly requested for removal (thanks @lazyhamster  for pointing this out).
    * This can be ignored if the `--recurse` flag is also added.
    * Should let the user know that the `--recurse` flag is needed in order to proceed.

Another idea:

* Should imply `--dry-run` and only print the removal plan. Require `--no-dry-run` explicitly to actually remove packages.

But I'm not sure if the last one offers a good UX.

**Comment by IIFE** [comment]
@vicroms Thanks for the reply.

I have done the implementation but hadn't got the chance to create a PR. Will revisit my implementation and create a PR when I get a chance.

Watch this space.

Thanks.

**Comment by diablodale** [comment]
Yes, this is strange behavior and due to a broken mental model for vcpkg. And that errant model leads to a orphanage and/or dependency tree with holes. @PhoebeHui writes...

> @IIFE, this is intended behavior, if you remove boost, then adding --recurse option will remove all ports(eg: cpprestsdk)that dependent to boost.

In that scenario, the `vcpkg remove boost` will result in a broken vcpkg library tree. Why?

1. cpprestsdk depends on boost being installed
2. `vcpkg remove boost` removes boost from the folders
3. Now any use of cpprestsdk has a high chance of failure depending on how cpprestsdk and boost were built (static, dynamic, etc.)

As I see it, `vcpkg remove boost` should immediately fail. vcpkg should forbid any removal of packages who have other packages that depend on them. 

As I have read and somewhat seen myself, the `vcpkg remove --recurse` doesn't recurse *down* into children. It recurses *up* into parents. It recurses to parent dependers, not child *dependencies*.

This conflicts with the mental model of `vcpkg install`. Why? First, `install` and `remove` are a matched pair. `vcpkg install` installs a package and its *dependencies*. Therefore, the mental model is that `vcpkg remove` would remove a package and its *dependencies*. But it doesn't.

| vcpkg action | who is affected |
| --- | --- |
| install | dependencies ✔|
| remove | dependers ❌| 

Since the mental model is not consistent in vcpkg, this causes the confusion of the OP (and me). `vcpkg remove` results in a broken dependency tree since the dependers are still on the hard drive. And leads to manual work to cleanup orphan *dependencies* that were not removed.

This then puts into question the option `--recurse`. It is a curious option. If you *don't* use it, then the vcpkg tree will have dependency holes since the parent depender's dependency was removed. If you *do* use it, then there won't be holes for the immediate port, but if it has any dependers (to which `--recurse` recurses ?infinitely?), then when the chain of dependers (and their dependers...) are removed it will leave all the dependencies of those dependers as orphans. Creating not holes...instead an broad orphanage. 😵

At my current level of thinking through this problem, it is unsafe to use `vcpkg remove`. It is additionally unsafe to use `vcpkg remove --recurse` on any package that has a dependency. Not following those two rules will result in a vcpkg folder with either a shotgun of dependency holes or an orphanage.

On the plus side, nobody dies here. No production-stage app crashes. Dependencies can be rebuilt when errors occur. And wasted hard drive space is cheap.

I reproduced this broken model with the opencv4 port. When I `vcpkg remove opencv4:triplet`, it left many orphaned packages including (hdf5, opengl, szip, gflags, glog, etc.). In https://github.com/microsoft/vcpkg/blob/master/ports/opencv4/CONTROL you can see all the dependencies. `vcpkg remove --recurse` only removed opencv4 and its features. It removed no dependencies.

Dependencies, dependers, and a ref count for each are challenges that tools like `apt` and `yum` have been managing for years. For example, `apt` tracks when *dependencies* are not longer needed and prompts the admin; who can then purge them with `apt autoremove`. There can be learning from these seasoned tools to address the broken vcpkg mental model and look forward to possible resolutions.

**Comment by IIFE** [comment]
@diablodale Pretty much summed up my experience.

The way I decided to use `vcpkg` going forward is through the `export` option. If I need a library, I will install it and export it to a raw folder. I will then integrate my project with the library by importing the `vcpkg.targets` file. This makes it much easier to integrate libraries without having to worry about conflicts that will arise with other branches that may get built on the same machine. Using `nuget` export just leads to headache as `nuget` is really not meant for C++. Also I find it easier to remove/update packages this way, as it's a matter of removing the exported folder or updating it. `vcpkg` simply becomes a convenient tool to download and build libraries, as well as having a useful MSBuild target script to integrate libraries.

The only issue I found with the above approach is that the MSuild macro (`VcpkgRoot`) that `vcpkg` uses in the `vcpkg.targets` file is the same for every exported package (see #3954). This means when you import multiple packages in the same project, they will conflict and only one of them will win. My fix for this issue is a post-processing script that modifies the `vcpkg.targets` of the exported package to append the package name to `VcpkgRoot` element name.

To summarise, the whole `remove` command is not reliable. Using `vcpkg` with global integration leads to issues if multiple branches are built on the same machine (dev or build machines). Exporting to `nuget` causes more headache that its worth. Best solution is to export to `raw` and import the `vcpkg.targets` file.

**Comment by cbezault** [comment]
If what @diablodale says is true generally this is a bug. `vcpkg remove --recurse port_a` should remove all ports which depend on `port_a` as well as `port_a`. Boost on the other hand could be a weird special case. The "boost" port is simply a convenience port for installing all the boost ports. No other port probably depends on `boost`. They almost certainly depend on pieces of boost like `boost-math` etc. Therefore `vcpkg remove boost` will just remove the convenience port but not any of the actual binaries.

Edit: Woops completely misread your comment. The behavior you're looking for could be implemented and I think such a PR might be accepted.

**Comment by diablodale** [comment]
@cbezault , please consider what your typed in your 2nd sentence and my analysis post above.
The Win32 binary `vcpkg` does that + leaves an orphanage. *That's* the issue.
This issue is not about a coding bug. This issue is about a design/Ux flaw. Someone spec'd a design model which doesn't match the mental model of users with the install/remove pair. While also creating situations that leave dependency holes and/or orphanages.

**Comment by diablodale** [comment]
This is actually issue https://github.com/microsoft/vcpkg/issues/8220. What more discussion is needed here? As I wrote above, this is a design/Ux that should be considered by the core Microsoft team.
